framework/infrastructure/traffic_director.py

```diff
@@ -612,30 +631,56 @@ def create_target_proxy(self):
             target_proxy_type,
             self.url_map.name,
         )
-        self.target_proxy = create_proxy_fn(name, self.url_map)
 
-    def delete_target_grpc_proxy(self, force=False):
+        if ipv6:
+            self.target_proxy_v6 = create_proxy_fn(name, self.url_map)
+        else:
+            self.target_proxy = create_proxy_fn(name, self.url_map)
+
+    def delete_target_grpc_proxy(self, force=False, ipv6=False):
         if force:
-            name = self.make_resource_name(self.TARGET_PROXY_NAME)
+            if ipv6:
+                name = self.make_resource_name(self.TARGET_PROXY_NAME_IPV6)
+            else:
+                name = self.make_resource_name(self.TARGET_PROXY_NAME)
+        elif ipv6:
+            if self.target_proxy_v6:
+                name = self.target_proxy_v6.name
         elif self.target_proxy:
             name = self.target_proxy.name
         else:
             return
+
         logger.info('Deleting Target GRPC proxy "%s"', name)
         self.compute.delete_target_grpc_proxy(name)
-        self.target_proxy = None
+        if ipv6:
+            self.target_proxy_v6 = None
+        else:
+            self.target_proxy = None
+
         self.target_proxy_is_http = False
 
-    def delete_target_http_proxy(self, force=False):
+    def delete_target_http_proxy(self, force=False, ipv6=False):
         if force:
-            name = self.make_resource_name(self.TARGET_PROXY_NAME)
+            if ipv6:
+                name = self.make_resource_name(self.TARGET_PROXY_NAME_IPV6)
+            else:
+                name = self.make_resource_name(self.TARGET_PROXY_NAME)
```

Comment:
Branching in this method gets pretty confusing for no good reason. We should refactor this to be similar break it up similar to the firewall rules, see `delete_firewall_rules`, `delete_firewall_rule`, `delete_firewall_rule_ipv6`

Reply:
I see. I was trying to avoid much code duplication but after combining dualstack it did look a bit complex. Refactored all creations and deletions to separate ipv6 functions.